### PR TITLE
Move UI stuff for UserVariables (global variables) from sdk to src and do not block on batch mode

### PR DESCRIPTION
**Pull request by bluehazzard**
This is part of my work to make codeblocks more usable in automated build environment.
One case where this was a problem is, when a global user variable was not defined. This popped up a info window and the New Variable dialogue. On a batch build on a headless machine this blocks the build without information for the user.

The idea behind this rework is to split UI stuff and SDK stuff. Move the UI stuff to src and define a minimal UI (without GUI) for batch builds.

I have defined a virtual base class UserVarManagerUI and derived two classes:
1) UserVarManagerDefaultUI -> this does basically nothing but output a warning message instead of a info window. I have tested this and it does not output anything to the console. We should probably simply do a cout in this
2) UserVarManagerGUI -> Full GUI for normal running

No big functional changes except if cb is run in batch mode no GUI for the UserVariables is generated

Question i hope to get answers here:
1) Can we use cout to display information to the console
2) Is this approach with pure virtual class + 2 derived classes ok?
3) Are the static const variables for the UserVariableManager ok?

thank you

### CODE REVIEW

File: src/include/uservarmanager.h [hunk 1]
```diff
@@ -40,6 +55,19 @@ class DLLIMPORT UserVariableManager : public Mgr<UserVariableManager>
...
+        static const wxChar   cSlash;
+        static const wxString cSlashBase;
+        static const wxString cInclude;
+        static const wxString cLib;
+        static const wxString cObj;
+        static const wxString cBin;
+        static const wxString cCflags;
+        static const wxString cLflags;
+        static const wxString cSets;
+        static const wxArrayString builtinMembers;
```
Comment (obfuscated) [inline src/include/uservarmanager.h:70]: Why are these made public? I don't see them using them. If you use them for something add a block doxygen comment stating the reason.

File: src/include/uservarmanager.h [hunk 2]
```diff
@@ -14,6 +14,15 @@
     #include "globals.h"
 #endif
 
+class UserVarManagerUI
```
Comment (obfuscated) [inline src/include/uservarmanager.h:17]: Doxygen comments depicting the interface of these.

File: src/include/uservarmanager.h [hunk 3]
```diff
@@ -24,8 +33,14 @@ class DLLIMPORT UserVariableManager : public Mgr<UserVariableManager>
         wxString        m_ActiveSet;
         wxArrayString   m_Preempted;
 
+        UserVarManagerUI* m_ui;
```
Comment (obfuscated) [inline src/include/uservarmanager.h:36]: std::unique_ptr please.
And get your free space act together. Why are there two empty lines.

File: src/sdk/uservarmanager.cpp [hunk 4]
```diff
@@ -174,11 +117,11 @@ wxString UserVariableManager::Replace(const wxString& variable)
...
                          "the global compiler variable \"%s\".\n\n"
                          "Please define it."), package.wx_str());
-            InfoWindow::Display(_("Global Compiler Variables"), msg , 8000, 1000);
-            UsrGlblMgrEditDialog d;
-            d.AddVar(package);
-            PlaceWindow(&d);
-            d.ShowModal();
+
+            m_ui->DisplayInfoWindow(_("Global Compiler Variables"), msg);
+            wxArrayString ar;
```
Comment (obfuscated) [inline src/sdk/uservarmanager.cpp:122]: Prefer std::vector in new code. I think.

File: src/sdk/uservarmanager.cpp [hunk 5]
```diff
@@ -260,19 +199,24 @@ void UserVariableManager::Arrogate()
                      "%s\n\n"
                      "Please define them."), peList.wx_str());
 
-    PlaceWindow(&d);
+    m_ui->DisplayInfoWindow(_("Global Compiler Variables"), msg);
```
Comment (obfuscated) [inline src/sdk/uservarmanager.cpp:202]: Why is this call here? Why don't you put it in the implementation of the OpenEditWindow call?

File: src/sdk/uservarmanager.cpp [hunk 6]
```diff
@@ -260,19 +199,24 @@ void UserVariableManager::Arrogate()
...
     m_Preempted.Clear();
-    InfoWindow::Display(_("Global Compiler Variables"), msg , 8000 + 800*m_Preempted.GetCount(), 100);
-
-    d.ShowModal();
 }
 
 UserVariableManager::UserVariableManager()
 {
     m_CfgMan = Manager::Get()->GetConfigManager(_T("gcv"));
+    m_ui =  new UserVarManagerDefaultUI(); //new UserVarManagerGUI(); //new UserVarManagerDefaultUI();
```
Comment (obfuscated) [inline src/sdk/uservarmanager.cpp:210]: No commented code please. No extra space after the space... More focus.

File: src/src/app.cpp [hunk 7]
```diff
@@ -730,6 +731,9 @@ bool CodeBlocksApp::OnInit()
             }
         }
 
+        if (!m_Batch)
+            Manager::Get()->GetUserVariableManager()->SetUI(new UserVarManagerGUI());
```
Comment (obfuscated) [inline src/src/app.cpp:735]: I think this should be something like `SetUI(!m_Batch? new UserVarManagerGUI : new UserVarManagerGUIBatch);`

File: src/src/uservardlgs.cpp [hunk 8]
```diff
@@ -0,0 +1,538 @@
...
+void GetUserVariableDialog::Load()
+{
+    if (m_treectrl == nullptr)
+        return;
+
+    m_treectrl->DeleteAllItems();
+
+    ConfigManager* CfgMan = Manager::Get()->GetConfigManager(wxT("gcv"));
+    const wxString &ActiveSet = Manager::Get()->GetConfigManager(wxT("gcv"))->Read(wxT("/active"));
+    wxArrayString vars = CfgMan->EnumerateSubPaths(UserVariableManager::cSets + ActiveSet + wxT("/"));
```
Comment (obfuscated) [inline src/src/uservardlgs.cpp:93]: This is an implementation detail of the UserVarManager, so it is something that shouldn't escape its implementation, but should be hidden behind an interface. But I suppose it is not the time to do it.

File: src/src/uservardlgs.h [hunk 9]
```diff
@@ -0,0 +1,117 @@
...
+    void UpdateChoices();
+    void AddVar(const wxString& var);
+    void Sanitise(wxString& s);
+
+    DECLARE_EVENT_TABLE()
+
+public:
+    UsrGlblMgrEditDialog(const wxString& var = wxEmptyString);
+    friend class UserVarManagerGUI;
+};
```
Comment (obfuscated) [inline src/src/uservardlgs.h:105]: These are implementation details of the UserVarManagerGUI, so they don't belong to the .h file. Just move them in the .cpp.

File: src/src/uservardlgs.h [hunk 10]
```diff
@@ -0,0 +1,117 @@
...
+
+
+class UserVarManagerGUI : public UserVarManagerUI
+{
+    public:
```
Comment (obfuscated) [inline src/src/uservardlgs.h:110]: Incorrect indent.

File: src/src/uservardlgs.h [hunk 11]
```diff
@@ -0,0 +1,117 @@
...
+    void DisplayInfoWindow(const wxString &title, const wxString &msg);
+    void OpenEditWindow(const wxArrayString &var);
+    wxString GetVariable(wxWindow* parent, const wxString &old);
```
Comment (obfuscated) [inline src/src/uservardlgs.h:113]: Missing override!

File: src/src/uservardlgs.h [hunk 12]
```diff
@@ -0,0 +1,117 @@
...
+};
+
+
```
Comment (obfuscated) [inline src/src/uservardlgs.h:116]: Extra new line.

File: src/src/Makefile.am [hunk 13]
```diff
@@ -63,6 +63,7 @@ codeblocks_SOURCES = app.cpp \
 			startherepage.cpp \
 			switcherdlg.cpp \
 			threadsdlg.cpp \
+			uservardlgs.cpp \
```
Comment (obfuscated) [inline src/src/Makefile.am:66]: What about the XRC file moves? Are they correctly incorporate in the autotools build? Using make dist and then listing the generated archive is a good way to check.

File: src/sdk/uservarmanager.cpp [hunk 14]
```diff
@@ -63,88 +63,31 @@ const wxChar *bim[] =
     _T("cflags"),
     _T("lflags")
 };
-const wxArrayString builtinMembers((size_t) 7, bim);
+const wxArrayString UserVariableManager::builtinMembers((size_t) 7, bim);
 
-class GetUserVariableDialog : public wxScrollingDialog
+class UserVarManagerDefaultUI : public UserVarManagerUI
```
Comment (obfuscated) [inline src/sdk/uservarmanager.cpp:68]: The name of this doesn't designate what it actually does...

File: src/include/uservarmanager.h [hunk 15]
```diff
@@ -14,6 +14,15 @@
     #include "globals.h"
 #endif
```
Comment (obfuscated) [inline src/include/uservarmanager.h:15]: Changes to this file means that you have to raise the version number in cbplugin.h. Do it as a last step before push.

**Comment by obfuscated** [comment]
Last few days I've made several changes to the UserVarManager is this branch incorporating them correctly? If not you'll introduce regressions.

I'm talking about these two: 81382c3fa4a848eed0d4cd4674cfeca49a841291 b9fdf36cd0806afbd2b2ca6e2386cd0ca6984053

File: src/sdk/uservarmanager.cpp [hunk 16]
```diff
@@ -63,88 +63,31 @@ const wxChar *bim[] =
...
-    wxString m_SelectedVar;
-    wxString m_old;
+    void DisplayInfoWindow(const wxString &title,const wxString &msg) override
+    {
+        Manager::Get()->GetLogManager()->LogWarning(msg);
+    }
 
-    DECLARE_EVENT_TABLE()
+    void OpenEditWindow(const wxArrayString &var) override { };
+    wxString GetVariable(wxWindow* parent, const wxString &old) override { return ""; };
```
Comment (obfuscated) [inline src/sdk/uservarmanager.cpp:77]: wxString() is more efficient than ""!

**Comment by obfuscated** [comment]
> The idea behind this rework is to split UI stuff and SDK stuff. Move the UI stuff to src and define a minimal UI (without GUI) for batch builds.

I'm not sure this is really great approach. I did something similar for the ProjectManager, but the code there is bigger and more involved. If I was fixing that problem for real I wouldn't do it with a UI interface class. I've done it like this because I wanted to solve a particular problem in a fast manner.

And here I'm not really sure you've fixed the actual problem. To me it seems that the actual problem is that we need a way to define global variables from the command line. If I were you I'd add this command line stuff and then I'd abort the build any time a variable is requested which is not defined (either in the config or on the command line). And I'll print an error message in the build log.
With the current approach you're forcing people to write/copy the config file on the build machines, which is not an easy task and it is error prone. 

Also if the variable which is not defined is controlling some feature which doesn't fail the build it might be really hard to detect if you don't inspect the log every time you make changes to the build/machines/etc, so it will lead to silently broken builds. And this is never a good thing in a build system.

Sorry if I've miss something in the implementation and the above is not really correct.

**Comment by bluehazzard** [comment]
> And here I'm not really sure you've fixed the actual problem. To me it seems that the actual problem is that we need a way to define global variables from the command line. If I were you I'd add this command line stuff and then I'd abort the build any time a variable is requested which is not defined (either in the config or on the command line). And I'll print an error message in the build log.
With the current approach you're forcing people to write/copy the config file on the build machines, which is not an easy task and it is error prone.
> Also if the variable which is not defined is controlling some feature which doesn't fail the build it might be really hard to detect if you don't inspect the log every time you make changes to the build/machines/etc, so it will lead to silently broken builds. And this is never a good thing in a build system.

You are right! And i was thinking about this things... My next planed step were to:
1) Introduce command line parameter for variables, that can override defined ones
2) Rework the UI to make it similar like CMake, so we can add descriptions and default values for global variables on a project level, also defining new variables is more intuitive, the current UI is terrible in this regard....
3) Scan Project files for global variables syntax on saving of the project and storing them separately in the project with the possibility to add description and default value

With this pull request i try to solve the concrete problem that batch builds fail silently if a user variable is not defined and move UI code out of the sdk library. I think the UI Interface is a universal solution.

But anyway i will investigate if there is a cleaner way to make this...

**Comment by obfuscated** [comment]
> With this pull request i try to solve the concrete problem that batch builds *fail silently* if a user variable is not defined and move UI code out of the sdk library. 

But then you fail to do it, because builds could still fail silently if the variable is not defined in the settings.
As a minimum I think you should return `<invalid>` `<not defined in the settings>` or something similar when a variable is requested and the UI should be shown. And you should do this no matter what you decide to do gui interface or just plain fail-the-build.

The fail-the-build solution is quite a lot shorter probably and it wouldn't cause problems when you start to do the cmake stuff later, because the code would be a lot less.

> I think the UI Interface is a universal solution.

What do you mean by universal?

p.s. Keep in mind that the cmake stuff is complicated and the cmake people have failed to do it properly. At work we have quite a complicated code to workaround what should have been available in cmake by default.

**Comment by bluehazzard** [comment]
> But then you fail to do it, because builds could still fail silently if the variable is not defined in the settings.
As a minimum I think you should return <invalid> <not defined in the settings> or something similar when a variable is requested and the UI should be shown. And you should do this no matter what you decide to do gui interface or just plain fail-the-build.

it would not fail silently, if i put a cout in the `UserVarManagerUI::DisplayInfoWindow` function it would give at least a message in the console that started codeblocks.... Not ideal but better as the current situation...

> As a minimum I think you should return <invalid> <not defined in the settings> or something similar when a variable is requested and the UI should be shown.

> p.s. Keep in mind that the cmake stuff is complicated and the cmake people have failed to do it properly. At work we have quite a complicated code to workaround what should have been available in cmake by default.

i talk only about the UI with | name | value | description | grid... 

Agree!

I looked a bit in the code and 99% of the UserVariableManager::Replace function is called in the macros manager.
As you say, the error should be logged in the build log, so how do we propagate the error from the UserVariableManager, through the Macros Manager to the compiler plugin?
* Are exception a possibility (probably the easiest solution, but hell, not the nicest)?
* Better rewrite the whole code and let MacroManager::ReplaceMacros return some kind of result list:
```
class ReplaceResult {
   wxString name;
   wxString errorMessage;
}

std::vector<ReplaceResult> UserVariableManager::Replace(wxString& buffer);
std::vector<ReplaceResult> MacrosManager::ReplaceMacros(wxString& buffer, const ProjectBuildTarget* target, bool subrequest);
```
?
Suggestions?

File: src/include/uservarmanager.h (hunk 2 above)
Comment (bluehazzard) [reply src/include/uservarmanager.h:19]: in a6f3a04

File: src/sdk/uservarmanager.cpp (hunk 14 above)
Comment (bluehazzard) [reply src/sdk/uservarmanager.cpp]: Name changed in e9f17b2

File: src/sdk/uservarmanager.cpp (hunk 6 above)
Comment (bluehazzard) [reply src/sdk/uservarmanager.cpp]: fixed in bf2875c

File: src/include/uservarmanager.h (hunk 3 above)
Comment (bluehazzard) [reply src/include/uservarmanager.h]: fixed in bf2875c

File: src/sdk/uservarmanager.cpp (hunk 4 above)
Comment (bluehazzard) [reply src/sdk/uservarmanager.cpp]: i replaced the wxArrayString with a std::set<wxString> in d88ee6c96

File: src/src/uservardlgs.h (hunk 11 above)
Comment (bluehazzard) [reply src/src/uservardlgs.h]: in 3fb97bb

File: src/src/uservardlgs.h (hunk 10 above)
Comment (bluehazzard) [reply src/src/uservardlgs.h:109]: in e58aa66c6

File: src/src/uservardlgs.h (hunk 12 above)
Comment (bluehazzard) [reply src/src/uservardlgs.h]: not anymore

**Comment by obfuscated** [comment]
There is nothing more I could do on this review. I've told you that you're overcomplicating it. If this is how you want to do it then I'm fine.

**Pull request closed without merging**